internal/provider/resource_ssh_key.go

```diff
@@ -5,184 +5,312 @@ import (
-			"model": {
+func (s *sshKeyResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Resource representing an SSH key.",
+		Attributes: map[string]schema.Attribute{
+			"model": schema.StringAttribute{
 				Description: "The name of the model to operate in.",
-				Type:        schema.TypeString,
 				Required:    true,
 			},
-			"payload": {
+			"payload": schema.StringAttribute{
 				Description: "SSH key payload.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				Sensitive:   true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 			},
 		},
 	}
 }
 
-func sshKeyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (s *sshKeyResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "ssh_key", "create")
+		return
+	}
 
-	var diags diag.Diagnostics
+	var plan sshKeyResourceModel
 
-	modelName := d.Get("model").(string)
-	modelUUID, err := client.Models.ResolveModelUUID(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	payload := d.Get("payload").(string)
-
+	payload := plan.Payload.ValueString()
 	user := utils.GetUserFromSSHKey(payload)
 	if user == "" {
-		return diag.Errorf("malformed SSH key, user not found")
+		resp.Diagnostics.AddError("Provider Error", fmt.Sprintf("malformed SSH key : %q", payload))
+		return
+	}
+
+	modelName := plan.ModelName.ValueString()
+	modelUUID, err := s.client.Models.ResolveModelUUID(modelName)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
 	}
 
-	err = client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
+	err = s.client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
 		ModelName: modelName,
 		ModelUUID: modelUUID,
 		Payload:   payload,
 	})
-
 	if err != nil {
-		diags = append(diags, diag.FromErr(err)...)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create ssh_key, got error %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("created ssh_key for: %q", user))
 
-	d.SetId(fmt.Sprintf("sshkey:%s:%s", modelName, user))
+	plan.ID = types.StringValue(newSSHKeyID(modelName, user))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
 
-	return diags
+func newSSHKeyID(modelName string, user string) string {
```

Comment:
because it's a helper function to make an ID, doesn't necessarily need to know about the `types` (i.e. it's up to you to use the result in whichever type you want).

https://github.com/juju/terraform-provider-juju/blob/ca7a9e73c9fd8da513caedb02be4c611f2b24299/internal/provider/resource_machine.go#L431